File: labs-opensource/M02-tokens-context-limits/solution/token_timer.js
/**
 * M02 Lab - Step 3: TokenTimer Benchmark — SOLUTION
 * ==================================================
 * Run: node token_timer.js
 */

import OpenAI from "openai";

class TokenTimer {
  constructor(model = "mistral") {
    this.model = model;
    this.client = new OpenAI({ baseURL: "http://localhost:11434/v1", apiKey: "ollama" });
  }

  async run(messages, maxTokens = 256) {
    const tStart = performance.now();
    let response;
    try {
      response = await this.client.chat.completions.create({
        model: this.model,
        messages,
        max_tokens: maxTokens,
      });
    } catch (err) {
      throw new Error(`Ollama request failed: ${err.message} (is Ollama running? ollama serve)`);
    }

    const elapsed = (performance.now() - tStart) / 1000; // seconds
    const usage = response.usage;
    const content = response.choices[0].message.content ?? "";
    const decodeTps = elapsed > 0 ? usage.completion_tokens / elapsed : 0;

    return {
      promptTokens: usage.prompt_tokens,
      completionTokens: usage.completion_tokens,
      totalTokens: usage.total_tokens,
      elapsedSeconds: elapsed,
      decodeTokensPerSec: decodeTps,
      content,
    };
  }
}

// ── Benchmark over three prompt sizes ──
function makePrompt(approxTokens) {
  const padding = "benchmark ".repeat(Math.floor(approxTokens / 2));
  return [
    { role: "system", content: "You are a helpful assistant. Answer briefly." },
    { role: "user", content: `Please summarize the following:\n\n${padding}\n\nSummarize in one sentence.` },
  ];
}

const timer = new TokenTimer("mistral");

console.log("Warming up...");
await timer.run([{ role: "user", content: "hi" }], 5);

console.log("\nBenchmark results:");
console.log(
  `${"Approx Input".padStart(14)} ${"Prompt Tok".padStart(10)} ${"Comp Tok".padStart(9)} ${"Elapsed".padStart(8)} ${"Decode tok/s".padStart(13)}`
);
console.log("-".repeat(60));

for (const target of [100, 500, 1_000]) {
  const msgs = makePrompt(target);
  const r = await timer.run(msgs, 64);
  console.log(
    `${(target + "->").padStart(14)} ` +
    `${String(r.promptTokens).padStart(9)} ` +
    `${String(r.completionTokens).padStart(9)} ` +
    `${r.elapsedSeconds.toFixed(2).padStart(7)}s ` +
    `${r.decodeTokensPerSec.toFixed(1).padStart(12)}`
  );
}

File: labs-opensource/M02-tokens-context-limits/solution/token_timer.py
"""
M02 Lab - Step 3: TokenTimer Benchmark — SOLUTION
==================================================
Run: python token_timer.py
"""

import time
from dataclasses import dataclass
from openai import OpenAI


@dataclass
class TimerResult:
    prompt_tokens: int
    completion_tokens: int
    total_tokens: int
    elapsed_seconds: float
    decode_tokens_per_sec: float
    content: str


class TokenTimer:
    """Wrap an Ollama call and report latency + throughput metrics."""

    def __init__(self, model: str = "mistral"):
        self.model = model
        self.client = OpenAI(base_url="http://localhost:11434/v1", api_key="ollama")

    def run(self, messages: list[dict], max_tokens: int = 256) -> TimerResult:
        """Send a chat request and return timing metrics."""
        t_start = time.perf_counter()
        try:
            response = self.client.chat.completions.create(
                model=self.model,
                messages=messages,
                max_tokens=max_tokens,
            )
        except Exception as e:
            raise RuntimeError(
                f"Ollama request failed: {e} (is Ollama running? ollama serve)"
            ) from e

        elapsed = time.perf_counter() - t_start
        usage = response.usage
        content = response.choices[0].message.content or ""

        # decode speed = output tokens / total time
        decode_tps = (usage.completion_tokens / elapsed) if elapsed > 0 else 0

        return TimerResult(
            prompt_tokens=usage.prompt_tokens,
            completion_tokens=usage.completion_tokens,
            total_tokens=usage.total_tokens,
            elapsed_seconds=elapsed,
            decode_tokens_per_sec=decode_tps,
            content=content,
        )


# ── Benchmark over three prompt sizes ──
def make_prompt(approx_tokens: int) -> list[dict]:
    """Generate a padding prompt of approximately the given token count."""
    padding = "benchmark " * (approx_tokens // 2)
    return [
        {"role": "system", "content": "You are a helpful assistant. Answer briefly."},
        {"role": "user", "content": f"Please summarize the following:\n\n{padding}\n\nSummarize in one sentence."},
    ]


def main():
    timer = TokenTimer(model="mistral")

    # Warm-up (avoids model load time polluting benchmarks)
    print("Warming up...")
    _ = timer.run([{"role": "user", "content": "hi"}], max_tokens=5)

    print("\nBenchmark results:")
    print(f"{'Approx Input':>14} {'Prompt Tok':>10} {'Comp Tok':>9} {'Elapsed':>8} {'Decode tok/s':>13}")
    print("-" * 60)

    for target in [100, 500, 1_000]:
        msgs = make_prompt(target)
        result = timer.run(msgs, max_tokens=64)
        print(
            f"{target:>12}-> "
            f"{result.prompt_tokens:>9} "
            f"{result.completion_tokens:>9} "
            f"{result.elapsed_seconds:>7.2f}s "
            f"{result.decode_tokens_per_sec:>12.1f}"
        )


if __name__ == "__main__":
    main()
